engine/class_modules/warlock/sc_warlock.cpp

```diff
@@ -55,6 +55,70 @@ struct drain_life_t : public warlock_spell_t
   }
 };  
 
+struct impending_catastrophe_dot_t : public warlock_spell_t
+{
+  impending_catastrophe_dot_t( warlock_t* p )
+    : warlock_spell_t( "impending_catastrophe_dot", p, p->find_spell( 322170 ) )
+  {
+    background = true;
+    may_miss   = false;
+    dual       = true;
+    aoe        = -1;
+    dot_duration *= 1 + p->conduit.catastrophic_origin.percent();
+  }
+};
+
+struct impending_catastrophe_impact_t : public warlock_spell_t
+{
+  impending_catastrophe_impact_t( warlock_t* p )
+    : warlock_spell_t( "impending_catastrophe_impact", p, p->find_spell( 322167 ) )
+  {
+    background = true;
+    may_miss   = false;
+    dual       = true;
+
+  }
+};
+
+struct impending_catastrophe_t : public warlock_spell_t
+{
+  action_t* impending_catastrophe_dot;
+  action_t* impending_catastrophe_impact;
+
+  impending_catastrophe_t( warlock_t* p, util::string_view options_str ) : 
+    warlock_spell_t( "impending_catastrophe", p, p->covenant.impending_catastrophe ),
+    impending_catastrophe_impact( new impending_catastrophe_impact_t(p) ),
+    impending_catastrophe_dot( new impending_catastrophe_dot_t( p ) )
+  {
+    parse_options( options_str );
+    travel_speed = 16;
+   
+    add_child( impending_catastrophe_impact );
+    add_child( impending_catastrophe_dot );
+  }
+
+  void impact( action_state_t* s ) override
+  {
+    auto targets = target_list();
+
+    warlock_spell_t::impact( s );
+
+    // This is where we would implement Curse of Weakness/Tongues
+    make_event<ground_aoe_event_t>( *sim, p(), ground_aoe_params_t()
+        .n_pulses(1)
+        .target( s->target )
+        .action( impending_catastrophe_dot ),
+        true );
+
```

Comment:
As mentioned in the comments on this PR, the event code should be changed into something that triggers the action directly on each target, either through aoe or as part of the cycle through the target list that occurs for the inflight damage